Team B. 무한슬라이딩 구현

Jeju-Boobby opened the pull request:
배민찬과 다르게 1번에 1개의 메뉴씩만 슬라이딩 되도록 구현하였습니다.

kkiluk committed:
테스트 코드를 정상화하고, 새로운 테스트 코드를 추가하다.

callback 형태에서 promise 형태로 변경된 ACResource.getData의 테스트 코드를 업데이트했습니다.
ACResource.cacheRecentData의 테스트 코드를 추가했습니다.

HannaJeon committed:
리뷰받은 내용 수정중

HannaJeon committed:
리뷰받은 내용 반영하다.

HannaJeon committed:
테스트 코드 버그 수정

kkiluk committed:
[FIX][REFACTORING] 자동완성의 포커스 아웃 기능을 수정하다.

search-bar에서 포커스가 아웃되었을 때의 버그를 수정했습니다.
리스트의 display를 담당하는 로직을 Responder에서 Renderer로 변경했습니다.

kkiluk committed:
[FIX] 최근검색어리스트의 클릭 이벤트를 수정하다.

리스트 아이템 안에 삭제 버튼이 존재하기 때문에, clickRemoveButton 메소드를 삭제하고, 해당 로직을 clickItem 메소드가 포함하도록 수정했습니다.

kkiluk committed:
[FIX] 공백 검색어 문제와 포커스아웃 버그를 수정하다.

검색입력창에 공백을 넣었을 때, 최근검색어에 저장되는 문제를 해결했습니다.
click 이벤트보다 focusout 이벤트가 먼저 콜되면서 발생하는 문제를 해결했습니다.

kkiluk committed:
[ADD] 무한슬라이딩 컴포넌트와 기능을 추가하다.

sliding-window & sliding-menu-list를 추가하고, 슬라이딩 로직을 구현했습니다.

Jeju-Boobby committed:
버튼에 이미지 삽입 및 레이아웃 조정

crongro reviewed (changes requested):
ACResponder 에 slide가 들어있는게 어색해보이는데요.
예를들어 이와 같은 기능이 페이지에 추가되면.. 점점 ACResponder가 비대해질 듯. 
AC라는 이름이...autoComplete로 알고 있기도 하고요.
그래서 별도 class로 구현해보면 좋겠고요.

메서드도 changeSearchText와 같은 녀석들은 너무 커요.
좀 작게 분리하면 좋겠어요.

slide완성 후 테스트코드 추가해보시고요.

kkiluk committed:
[REFACTORING] 무한슬라이딩을 담당하는 클래스를 생성하다.

무한슬라이딩의 View를 담당하는 InfiniteSlidingRenderer 클래스와 Controller를 담당하는 InfiniteSlidingResponder 클래스를 생성하고,
기존 AutoCompleteResponder에 포함되어있던 무한슬라이딩 기능을 각각 분리했습니다.

Jeju-Boobby committed:
[Added] 무한슬라이딩 2초간견으로 자동 슬라이딩되도록 구현

setInterval 함수를 이용해 2초마다 한번씩 오른쪽으로 슬라이딩 되도록
구현하였다. 수동으로 슬라이드를 한 후 Timer를 리셋시킨후 다시 작동하도록
기능을 구현하였다.

Jeju-Boobby committed:
[Added] 무한슬라이딩 테스트 구현하다

버튼 클릭 또는 transition end 이벤트가 들어왔을때 예상된 값이 나오는지
확인하는 테스트를 구현하다

crongro commented on src/app.js:
```diff
@@ -330,6 +331,12 @@ class InfiniteSlidingResponder {
+
+	resetInterval() {
+		clearInterval(this.autoInterval);
+		this.autoInterval = setInterval(this.startRightSliding.bind(this), 2000);
```
setInterval 보다는 requestAnimationframe이 더 많이쓰이는데요, 
이 두개의 함수의 차이를 알아야하고, 
그전에 setInterval의 문제점을 이해하는 게 좋습니다. 일정한 간격으로 실제 실행되지 않는 문제가 있어요.
